### PR TITLE
Add PlaylistHasItem method to detect duplicate playlist item

### LINKED ISSUE #1914: Playlist allows for duplicates
**Describe the bug**
If you add a media file to a playlist, there is no warning to let you know that it already exists in the playlist. It seems strange to have a need for duplicates in the first place and is probably not something that most users would want anyways.

**To Reproduce**
1. Go to a song or video item in Jellyfin
2. Click the + to add it to a playlist
3. Nothing will warn you if the item is already in the playlist

**Expected behavior**
After talking to dkanada on IRC, he suggested that we not make any client side changes and instead print an error message to the server logs if a duplicate item is found.

**Logs**
Not relevant in this case

**Screenshots**
N/A

**System (please complete the following information):**
N/A This would be a server side change for all operating systems

**Additional context**
I plan on making a commit with a new method in Emby.Server.Implementation\Playlists\PlaylistManager.cs  that does a duplicate check

### PR DESCRIPTION
**Pull request by SegiH**
**Changes**
Add PlaylistHasItem method to detect duplicate playlist item before add a track to a playlist

**Issues**
Fixes #1914

### CODE REVIEW

File: Emby.Server.Implementations/Playlists/PlaylistManager.cs [hunk 1]
```diff
@@ -227,6 +256,7 @@ private void AddToPlaylistInternal(string playlistId, IEnumerable<Guid> itemIds,
                 RefreshPriority.High);
         }
 
+
```
Comment (dkanada) [inline Emby.Server.Implementations/Playlists/PlaylistManager.cs]: only one newline here

File: CONTRIBUTORS.md [hunk 2]
```diff
@@ -30,10 +30,12 @@
  - [Khinenw](https://github.com/HelloWorld017)
  - [fhriley](https://github.com/fhriley)
  - [nevado](https://github.com/nevado)
+ - [SegiH](https://github.com/SegiH)
  - [mark-monteiro](https://github.com/mark-monteiro)
  - [ullmie02](https://github.com/ullmie02)
  - [pR0Ps](https://github.com/pR0Ps)
 
+
```
Comment (dkanada) [inline CONTRIBUTORS.md]: only one newline here

File: Emby.Server.Implementations/Playlists/PlaylistManager.cs [hunk 3]
```diff
@@ -201,13 +223,20 @@ private void AddToPlaylistInternal(string playlistId, IEnumerable<Guid> itemIds,
...
             foreach (var item in items)
             {
-                list.Add(LinkedChild.Create(item));
+                if (PlaylistHasItem(playlistId, item.Id.ToString()) == false)
+                {
+                    list.Add(LinkedChild.Create(item));
+                }
+                else
+                {
+                    throw new ArgumentException("The Playlist " + playlist.Name + " already has the item " + item.Name + " with the item ID " + item.Id);
```
[the remark this answers is not in the document]
Comment (dkanada) [reply Emby.Server.Implementations/Playlists/PlaylistManager.cs]: The exception itself is logging but I'm not sure we want to throw an exception at all, maybe just a single log.

File: Emby.Server.Implementations/Playlists/PlaylistManager.cs [hunk 4]
```diff
@@ -177,6 +177,28 @@ private List<BaseItem> GetPlaylistItems(IEnumerable<Guid> itemIds, string playli
...
+            var playlist = _libraryManager.GetItemById(playlistId) as Playlist;
+
+            if (playlist == null)
+            {
+                throw new ArgumentException("No Playlist exists with the supplied Id");
+            }
+
+            var playlistItems = playlist.LinkedChildren;
+
+            foreach (var item in playlistItems)
```
Comment (cvium) [inline Emby.Server.Implementations/Playlists/PlaylistManager.cs]: Could also do `return playlistItems.Any(item => item.Id == addingItemId);`

File: Emby.Server.Implementations/Playlists/PlaylistManager.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Server.Implementations/Playlists/PlaylistManager.cs]: Don't throw an exception and state that duplicate items will be ignored.

[the remark this answers is not in the document]
Comment (Bond-009) [reply Emby.Server.Implementations/Playlists/PlaylistManager.cs]: Btw, what if `itemIds` itself contains duplicate items?

File: Emby.Server.Implementations/Playlists/PlaylistManager.cs [hunk 5]
```diff
@@ -204,7 +226,14 @@ private void AddToPlaylistInternal(string playlistId, IEnumerable<Guid> itemIds,
 
             foreach (var item in items)
```
Comment (mark-monteiro) [inline Emby.Server.Implementations/Playlists/PlaylistManager.cs]: I think it should be possible to filter out the duplicates in a much simpler way, something like this:
```
var existingIds = playlist.LinkedChildren.Select(c => Guid.Parse(c.Id))
var uniqueIds = itemIds.Distinct().Except(existingIds);
```
This is much more readable to me, and should also be faster than checking each id in a loop. The call to `Distinct()` also handles the case where `itemIds` contains duplicate entries

**Comment by joshuaboniface** [comment]
@SegiH Do you have any comments on the review feedback?

We could probably apply the changes requested, but I'd prefer to keep the discussion going.

**Comment by SegiH** [comment]
I understand exactly what cviums line of code does and seems like it'd be fine.

So if I understand it correctly,  if the user tries to add a duplicate it will silently fail and log a message to the server log right ? In other words, the end user won't see a message that they're trying to add a dupe.

Is that correct ?
Is that right ?

**Comment by dkanada** [comment]
@SegiH we should be able to merge #2497 soon which includes your name in the contributors file since it used this pull request as a base.

**Comment by SegiH** [comment]
Sounds good thanks

**Pull request closed without merging**